### PR TITLE
//homework

### PR DESCRIPTION
**Pull request by CoderOfTomorrow**
//Added currency 
//Add currency convertor if source currency and destiantion currency are different

### CODE REVIEW

File: Server/CurrencyManager.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            };
+
+            Currencies = CurrencyDb.Select(e => e.Name).ToList();
+        }
+
+        public static decimal CheckCurrency(decimal amount,string source_currency,string destination_currency)
+        {
+           
+            decimal destination_amount=0;
+            if (source_currency == destination_currency)
```
Comment (cristianciolacu) [inline Server/CurrencyManager.cs]: Please check the naming conventions mentioned in the previous call.

File: Server/CurrencyManager.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                return amount;
+            else
+            {
+                decimal source_value = CurrencyDb.FirstOrDefault(e => e.Name == source_currency).Value;
```
Comment (cristianciolacu) [inline Server/CurrencyManager.cs]: What does this do?

Comment (CoderOfTomorrow) [reply Server/CurrencyManager.cs]: So , my system is base on MDL as base currency (value = 1) , when sourceCurrency is different from destinationCurrency , we a lookink for sourceCurrency value and multiply sourceAmount with it , in this way we exchange all the amount in MDL then we a looking for destinationCurrency value to divied it and obtine the destinationAmount in his currency

File: Server/Controllers/TransactionController.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        [HttpGet]
+        [Route("madetransactions")]
+        public List<Transaction> GetMadeTransactions()
+        {
+            var transactionsList = new List<Transaction>();
+            var userId = userManager.GetUserId(User);
+            var idList = context.Users.Include(e => e.Wallets).FirstOrDefault(x => x.Id == userId).Wallets.Select(e => e.Id).ToList();
+            transactionsList = context.Transactions.Where(t => idList.Contains(t.SourceWalletId) && t.DestinationUserId != t.SourceUserId).ToList();
```
Comment (estratulat) [inline Server/Controllers/TransactionController.cs]: I don't get what this `t.DestinationUserId != t.SourceUserId` does

File: Server/Controllers/TransactionController.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        [HttpGet]
+        [Route("interntransactions")]
+        public List<Transaction> GetInternTransactions()
+        {
+            var transactionsList = new List<Transaction>();
+            var userId = userManager.GetUserId(User);
+            var idList = context.Users.Include(e => e.Wallets).FirstOrDefault(x => x.Id == userId).Wallets.Select(e => e.Id).ToList();
+            transactionsList = context.Transactions.Where(t => idList.Contains(t.SourceWalletId) && idList.Contains(t.DestinationWalletId)).ToList();
```
Comment (estratulat) [inline Server/Controllers/TransactionController.cs]: Does this looks exactly the same as `GetTransactions` or I'm missing something?

File: Server/Controllers/TransactionController.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        public TransactionController(ApplicationDbContext context, UserManager<ApplicationUser> userManager)
+        {
+            this.context = context;
+            this.userManager = userManager;
+        }
+
+        [HttpGet]
+        [Route("transactions")]
+        public List<Transaction> GetTransactions()
```
Comment (estratulat) [inline Server/Controllers/TransactionController.cs]: Repeating code is a bad practice. Better combine all these alike methods. For example make a switch parameter that determines how filtering behaves.

File: Client/Pages/Transactions.razor [hunk 6]
```diff
@@ -1,151 +1,140 @@
...
+        transactionsList = await HttpClient.GetFromJsonAsync<List<Transaction>>($"api/transaction/interntransactions");
+    }
+
+    public async Task CallMethod(string data)
+    {
+        //Type thisType = this.GetType();
+        //MethodInfo theMethod = thisType.GetMethod(data);
+        //theMethod.Invoke(this, null);
+
+        if (data == "LoadMadeTransactions")
```
Comment (estratulat) [inline Client/Pages/Transactions.razor]: wouldn't it be simpler if you pass data as a parameter?

File: Server/Controllers/TransactionController.cs (hunk 3 above)
Comment (CoderOfTomorrow) [reply Server/Controllers/TransactionController.cs]: In my system users can make a transfer from one of his wallet to another. When i do the sortation of outgoing or recived transfer i dont want there to be transfer between his own wallets.

File: Server/Controllers/TransactionController.cs (hunk 4 above)
Comment (CoderOfTomorrow) [reply Server/Controllers/TransactionController.cs]: Get transaction - for all transactions / GetMadeTransactions for outgoin transactions / GetRecivedTransactions for recived transactions and GetInternalTransactions for transactions between our own wallets

File: Server/Controllers/TransactionController.cs (hunk 3 above)
Comment (estratulat) [reply Server/Controllers/TransactionController.cs]: What if the user wants to see them? Do you have a specific case for this?

Comment (CoderOfTomorrow) [reply Server/Controllers/TransactionController.cs]: yes ^^

File: Server/Controllers/TransactionController.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            var transactionData = new TransactionHistoryData { 
+                TransactionsList = new List<TransactionDto>()
+            };
+            var transactionsList = new List<Transaction>();
+            var userId = userManager.GetUserId(User);
+            var idList = context.Users.Include(e => e.Wallets).FirstOrDefault(x => x.Id == userId).Wallets.Select(e => e.Id).ToList();
+
+            transactionsList = typeFilter switch
+            {
+                "Made" => context.Transactions.Where(t => idList.Contains(t.SourceWalletId) && t.DestinationUserId != t.SourceUserId).ToList(), //for outgoing transactions
```
Comment (TermanEmil) [inline Server/Controllers/TransactionController.cs:39]: Would an enum for typeFilter, sortFilter, orderFilter be appropriate?

File: Server/Controllers/TransactionController.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        [HttpGet]
+        [Route("{typeFilter}/{sortFilter}/{orderFilter}/{itemsPerPage}/{pageNumber}")]
+        public TransactionHistoryData GetTransactions(string typeFilter,string sortFilter,string orderFilter, int itemsPerPage, int pageNumber)
+        {
+            var transactionData = new TransactionHistoryData { 
+                TransactionsList = new List<TransactionDto>()
+            };
+            var transactionsList = new List<Transaction>();
+            var userId = userManager.GetUserId(User);
+            var idList = context.Users.Include(e => e.Wallets).FirstOrDefault(x => x.Id == userId).Wallets.Select(e => e.Id).ToList();
```
Comment (TermanEmil) [inline Server/Controllers/TransactionController.cs:35]: Would it make sense to have walletIds instead of IdList?

Comment (TermanEmil) [inline Server/Controllers/TransactionController.cs:35]: I believe if Wallet had an owner property (User), then we could do something like this:
`context.Wallets.Where(x => x.Owner.Id == userId).Select(x => x.Id).ToList()`
\+ your query will fail if FirstOrDefault returns null

Comment (CoderOfTomorrow) [reply Server/Controllers/TransactionController.cs:35]: In general i want to change the finding of transaction by wallets id to userid , beacause if a user will delet the wallet , he will not be able to see transaction made with that wallet . What do you think .

Comment (TermanEmil) [reply Server/Controllers/TransactionController.cs:35]: Good point

Comment (TermanEmil) [reply Server/Controllers/TransactionController.cs:35]: But here, I was thinking to rename idList to walletIds or something. To me, it's not quite clear what these Ids are for. I had to analyze the query first

Comment (CoderOfTomorrow) [reply Server/Controllers/TransactionController.cs:35]: yes , i need to work more on variables name.

**Pull request merged**